compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -287,35 +288,114 @@ impl<'thir, 'p, 'tcx> MatchVisitor<'thir, 'p, 'tcx> {
         }
     }
 
+    /// Inspects the match scrutinee expression to determine whether the place it evaluates to may
+    /// hold invalid data.
+    fn is_known_valid_scrutinee(&self, scrutinee: &Expr<'tcx>) -> bool {
+        use ExprKind::*;
+        match &scrutinee.kind {
+            // Both pointers and references can validly point to a place with invalid data.
```

Comment:
For references I'd say this is an open question, but we want to be conservative here.

Reply:
Ah indeed, I'll update the comment to say that. Does the rest of this function seem right to you?

Reply:
I don't quite understand how it is used, but it treats unions and raw ptr deref specially so that seems right. For reference deref I guess we have to see how we want to treat them in the end, but it seems good to be conservative -- and I assume "not known to be valid" is the conservative answer here.

Reply:
What I am missing is test coverage, is there any test for the case of a union with an uninhabited field, or raw pointers?

Reply:
> I assume "not known to be valid" is the conservative answer here

Yep

Here's the test cases: https://github.com/Nadrieril/rust/blob/c3df51a976dfa69f0f1869997fdf79516ba6afab/tests/ui/pattern/usefulness/empty-types.rs#L191-L271

Reply:
Ah, nice! The tests for references should also have the comment about this being undecided so the lint being conservative.

Feel free to r? me on the PR that adds the comments.